substratevm/DEBUGINFO.md

```diff
@@ -0,0 +1,127 @@
+Using the ptototype debug info feature
+--------------------------------------
+
+To add debug info to a generated native image add flag
+-H:+TrackNodeSourcePosition to the native image command line.
+
+  mx native-image -H:+TrackNodeSourcePosition Hello.java
+
+The resulting image should contain code (method) debug symbols.
+
+What is currently implemented
+-----------------------------
+
+The currently implemented features include:
+
+  - break points configured by file and line or by method name
+  - single stepping by line including both into and over function calls
+  - stack backtraces (not including frames detailing inlined code)
+
+Note that single stepping within a compiled method includes file and
+line number info for inlined code, including inlined Graal methods.
+So, gdb may switch files even though you are still in the same
+compiled method.
+
+Identifying the location of source code
+---------------------------------------
+
+In order for gdb to be able to locate the source files for your app
+methods, Graal methods and JDK runtime methods you need to provide gdb
+with a list of source root dirs a 'set directories' command:
+
+    (gdb) set directories /home/adinn/hello/src:/home/adinn/graal/sdk/src/org/graalvm.word/src:/home/adinn/graal/sdk/src/org.graalvm.options/src:...
```

Comment:
Instead of having a bash script that will get very complicated over time as part of building debuginfo produce a `sources` folder next to the binary (subdir of `Path` HostedOption) where you collect all the source files that the generated debuginfo refers to. This way the debugger only needs to know about a single source dir. Also keep in mind that in the Java ecosystem jar-files often have their corresponding zip or jar file that contain the sources. In Java its easy to aggregate all these files from source bundles into a single common directory location.

Reply:
... also for the Windows version you would need to provide a `*.bat` version of that script. Don't go down that path ...

Reply:
Yeah, I'll get onto that after I sort out using Path instead of String.

Reply:
I have just pushed a change to populate a sources tree in the working directory using source file lookups based on whatever sources  can be found in $JAVA_HOME/lib/src/zip for JDK classes ($JAVA_HOME/lib/src/zip for jdk11) or via the classpath for GraalVM & application classes. The cache manager triages classes based on package prefix and uses a lookup scheme appropriate to the identified target type: jdk, graalvm or application.

JDK classes are looked up in src.zip, prepending a module name to the lookup path if necessary

GraalVM classes are looked up relative to the classpath. If foo.jar is found on the class path then if foo.src.zip is found it is assumed to contain graalvm sources and added to the list of search roots. If a directory foo is found on the class path and foo/src or foo/src_gen exist then they are added to the list of search roots.

Application sources are looked up relative to the classpath. If foo.jar is found on the class path then if foo-sources.jar is found it is assumed to contain application sources and added to the list of search roots. If a directory foo/classes or foo/target/classes is found on the class path and foo/src exists it is added to the list of search roots. The current working directory is also added to the list of application search roots.

Files are cached in local dir sources under 3 distinct subdirectories jdk, graalvm and src (for JDK, GraalVM and application sources, respectively). So, in order to make this work with gdb that requires adding 3 entries to the source search path

    (gdb) set directories /path/to/pwd/sources/jdk:/path/to/pwd/sources/graalvm:/path/to/pwd/sources/src

The file names and paths used in the debug info are normally derived using the package name, substituted with the local file system separator, suffixed with the classfile source name. If the source name is missing then the unqualified class name is used. In the latter case for inner classes the name before the $ is used. Class names which start with $ (e.g. anon classes etc) are not mapped to files and so cannot have debug entries. Of course, paths for JDK classes are prefixed with the name of the module that the class belongs to.

There is one problem with this. Compiler classes which are referenced from NodeSourcePosition records are not present in jars on the classpath (well, not when running with jdk11). For example, class org.graalvm.compiler.replacements.amd64.AMD64StringLatin1Substitutions is referenced from a simple HelloWorld app but the corresponding source is not found. That's because these classes are made available via the platform classloader as jmod files. In consequence there is no way to locate a source for these classes.

There are actually sources for classes like this in src.zip. However, 1) they are almost always going to be out of date wrt the source used by the native image generator and 2) they are bundled into the src.zip with both under a path which includes not just their module name but also a secondary directory based on some initial segment of their package prefix.

I am not really sure how to make these classes available. It appears that all the source files needed are available under compiler/mxbuild/dists/jdk11/ i.e. the directory that contains the bundled jmod files (they seem to be mostly in graal.src.zip). Is there any way for the native image generator to identify the location of this dir from the build config info and search it for extra roots to add to the GraalVM search path?

Reply:
@olpaw Forgot to tag this for your attention

Reply:
That sounds great!
I should be able to take a closer look somewhere in the next two days or so.

Reply:
Thanks. I'm just updating the DEBUGINFO.md readme which will clarify the latest status quo.